### PR TITLE
Add Advanced Sync Rules for Notion connector

### PR DESCRIPTION
**Pull request by praveen-elastic**
## Part of https://github.com/elastic/connectors-py/issues/1958 

This PR adds Advanced Sync Rules for Notion connector.

#### Pre-Review Checklist
- [x] this PR has a meaningful title
- [x] this PR links to all relevant github issues that it fixes or partially addresses
- [x] if there is no GH issue, please create it. Each PR should have a link to an issue
- [x] this PR has a thorough description
- [x] Covered the changes with automated tests
- [x] Tested the changes locally
- [x] Added a label for each target release version (example: `v7.13.2`, `v7.14.0`, `v8.0.0`)
- [x] Considered corresponding documentation changes
- [ ] Contributed any configuration settings changes to the configuration reference
- [ ] if you added or changed Rich Configurable Fields for a Native Connector, you made a corresponding PR in [Kibana](https://github.com/elastic/kibana/blob/main/packages/kbn-search-connectors/types/native_connectors.ts)

#### Changes Requiring Extra Attention

- [ ] Security-related changes (encryption, TLS, SSRF, etc)
- [ ] New external service dependencies added.

## Release Note

Support added for advanced sync rules in Notion Connector.

### CODE REVIEW

File: tests/sources/test_notion.py [hunk 1]
```diff
@@ -548,3 +556,126 @@ async def test_fetch_children_recursively():
...
+                ]
+            },
+            SyncRuleValidationResult(
+                SyncRuleValidationResult.ADVANCED_RULES,
+                is_valid=False,
+                validation_message=ANY,
+            ),
+        ),
+    ],
+)
```
Comment (seanstory) [inline tests/sources/test_notion.py:668]: can you add examples where `is_valid=True`? Please be sure to capture the edge cases like:
- multiple searches
- different filter types
- illegal checkbox values
- etc

File: tests/sources/test_notion.py [hunk 2]
```diff
@@ -548,3 +556,126 @@ async def test_fetch_children_recursively():
...
+                "searches": [
+                    {
+                        "filter": {"value": "database"},
+                        "query": "Demo",
+                    }
+                ]
+            },
+            SyncRuleValidationResult(
+                SyncRuleValidationResult.ADVANCED_RULES,
+                is_valid=False,
```
Comment (seanstory) [inline tests/sources/test_notion.py:623]: Can you add comments to each of these to explain why each is False? This will help us maintain/troubleshoot down the road.

File: docs/sync-rules/NOTION.md [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+
+  "database_query_filters" : [
+    {
+    "database_id": "notion_database_id1",
+    "filter": ​​{
+		"property": "Task completed",
+	      "checkbox": {
+	        "equals": true
+	      }
+	    }
```
Comment (seanstory) [inline docs/sync-rules/NOTION.md:164]: formatting looks a little weird here

File: docs/sync-rules/NOTION.md [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+    }
+  ]
+}
+
+]
+```
+**Note** 
+For `database_query_filters`:
+    - `property` contains a string that matches the database column name.
+    - `checkbox` this is not constant depends on the column name property type it can be `date`,`files`,`formula`,`multi_select`,`number`,`people`,`phone_number`,`relation`,`rich_text`,`select`,`status`,`timestamp` or `ID`
```
Comment (seanstory) [inline docs/sync-rules/NOTION.md:174]: I don't think I saw this restriction captured in the JSON Schema. Why not add it?

File: connectors/sources/notion.py [hunk 5]
```diff
@@ -158,18 +163,138 @@ async def fetch_comments(self, block_id):
...
+        "additionalProperties": False,
+    }
+
+    SCHEMA = fastjsonschema.compile(definition=RULES_OBJECT_SCHEMA_DEFINITION)
+
+    def __init__(self, source):
+        self.source = source
+
+    async def validate(self, advanced_rules):
+        if len(advanced_rules) == 0:
```
Comment (seanstory) [inline connectors/sources/notion.py:229]: Please add some logging to this function, so that we'll be able to troubleshoot what went wrong if something goes wrong.

File: connectors/sources/notion.py [hunk 6]
```diff
@@ -443,3 +555,53 @@ async def get_docs(self, filtering=None):
                     block_id=block_id
                 ):
                     yield self._format_doc(comment), None
+
+    async def get_docs(self, filtering=None):
```
Comment (seanstory) [inline connectors/sources/notion.py:559]: Please add some logging so that it's clear which flow we're going down, and what queries are being executed against the notion client.

**Review by seanstory — COMMENTED** [review]
Largely looks good. I don't see any obvious bugs, but I would like to see some more observability and some tighter tests. Left comments to that effect.

Thanks!

**Comment by praveen-elastic** [comment]
Buildkite test this

File: docs/sync-rules/NOTION.md (hunk 3 above)
Comment (praveen-elastic) [reply docs/sync-rules/NOTION.md]: Fixed the formatting issues 👍

File: docs/sync-rules/NOTION.md (hunk 4 above)
Comment (praveen-elastic) [reply docs/sync-rules/NOTION.md]: Since these are the list of filters already listed on Notion's documentation I have removed them now from our docs and added the link to the Notion Documentation highlighting it supports Notion filter syntax. Also, it would be quite complex for handling all of them in JSON schema when compounded with `and` `or`, therefore we can't add them all in schema.

Comment (seanstory) [reply docs/sync-rules/NOTION.md]: I was just figuring we could use a [json schema enum](https://json-schema.org/understanding-json-schema/reference/enum) for `checkbox`, no?

**Review by seanstory — APPROVED** [review]
Looks good, with the one lingering question about JSON schema enum. If you've got a good answer for that though, I think this is ready to go in.

File: docs/sync-rules/NOTION.md (hunk 4 above)
Comment (praveen-elastic) [reply docs/sync-rules/NOTION.md]: [json schema enum](https://json-schema.org/understanding-json-schema/reference/enum) cannot be used here since it only works for defining allowed list of values for the keys known in advance, for example, in the below schema, enum is used to provide allowed list of values for `street_type`. However, in our case we want to define allowed list of keys for an object. 

Enum use-case:
```
{
  "type": "object",
  "properties": {
    "street_type": { "enum": ["Street", "Avenue", "Boulevard"] }
  },
}
```

Comment (seanstory) [reply docs/sync-rules/NOTION.md]: I thought that `checkbox` was always the key name, with an enum of potential values like [`date`,`files`,`formula`,`multi_select`,`number`,`people`,`phone_number`,`relation`,`rich_text`,`select`,`status`,`timestamp`, `ID`]

**Comment by praveen-elastic** [comment]
> Looks good, with the one lingering question about JSON schema enum. If you've got a good answer for that though, I think this is ready to go in.

Merging the PR since we can't use JSON enum schema for our case as discussed above.
